eppo_core/src/event_ingestion/delivery.rs

```diff
@@ -72,7 +72,8 @@ pub(super) async fn delivery(
 
         let BatchedMessage { batch, flush } = msg;
 
-        let mut result = event_delivery.deliver(batch).await;
+        let delivery = event_delivery.lock().await;
+        let mut result = delivery.deliver(batch).await;
```

Comment:
This is holding the lock for too long (until the end of the loop iteration). Use nesting to scope it better:
```suggestion
        let mut result = {
          let delivery = event_delivery.lock().await;
          delivery.deliver(batch).await
        };
```